httpbakery/agent/agent_test.go

```diff
@@ -149,97 +146,201 @@ func (s *agentSuite) TestLoginCookie(c *gc.C) {
 		expectKey:  &key.Public,
 	}, {
 		about:       "no cookie",
-		setCookie:   func(client *httpbakery.Client, u *url.URL) {},
+		setCookie:   func(req *http.Request) {},
 		expectError: "no agent-login cookie found",
 		expectCause: agent.ErrNoAgentLoginCookie,
 	}, {
 		about: "invalid base64 encoding",
-		setCookie: func(client *httpbakery.Client, u *url.URL) {
-			client.Jar.SetCookies(u, []*http.Cookie{{
+		setCookie: func(req *http.Request) {
+			req.AddCookie(&http.Cookie{
 				Name:  "agent-login",
 				Value: "x",
-			}})
+			})
 		},
 		expectError: "cannot decode cookie value: illegal base64 data at input byte 0",
 	}, {
 		about: "invalid JSON",
-		setCookie: func(client *httpbakery.Client, u *url.URL) {
-			client.Jar.SetCookies(u, []*http.Cookie{{
+		setCookie: func(req *http.Request) {
+			req.AddCookie(&http.Cookie{
 				Name:  "agent-login",
 				Value: base64.StdEncoding.EncodeToString([]byte("}")),
-			}})
+			})
 		},
 		expectError: "cannot unmarshal agent login: invalid character '}' looking for beginning of value",
 	}, {
 		about: "no username",
-		setCookie: func(client *httpbakery.Client, u *url.URL) {
-			agent.SetCookie(client.Jar, u, "", &key.Public)
+		setCookie: func(req *http.Request) {
+			agent.SetCookie(req, "", &key.Public)
 		},
 		expectError: "agent login has no user name",
 	}, {
 		about: "no public key",
-		setCookie: func(client *httpbakery.Client, u *url.URL) {
-			agent.SetCookie(client.Jar, u, "hello", nil)
+		setCookie: func(req *http.Request) {
+			agent.SetCookie(req, "bob", nil)
 		},
 		expectError: "agent login has no public key",
 	}}
-	var (
-		foundUser string
-		foundKey  *bakery.PublicKey
-		foundErr  error
-	)
-	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		foundUser, foundKey, foundErr = agent.LoginCookie(req)
-	}))
-	defer srv.Close()
-
-	srvURL, err := url.Parse(srv.URL)
-	c.Assert(err, gc.IsNil)
 
 	for i, test := range tests {
 		c.Logf("test %d: %s", i, test.about)
 
-		client := httpbakery.NewClient()
-		client.Key = key
-		test.setCookie(client, srvURL)
-
-		req, err := http.NewRequest("GET", srv.URL, nil)
-		c.Assert(err, gc.IsNil)
-		resp, err := client.Do(req)
+		req, err := http.NewRequest("GET", "", nil)
 		c.Assert(err, gc.IsNil)
-		c.Assert(resp.StatusCode, gc.Equals, http.StatusOK)
+		test.setCookie(req)
+		username, key, err := agent.LoginCookie(req)
+
 		if test.expectError != "" {
-			c.Assert(foundErr, gc.ErrorMatches, test.expectError)
+			c.Assert(err, gc.ErrorMatches, test.expectError)
 			if test.expectCause != nil {
-				c.Assert(errgo.Cause(foundErr), gc.Equals, test.expectCause)
+				c.Assert(errgo.Cause(err), gc.Equals, test.expectCause)
 			}
 			continue
 		}
-		c.Assert(foundUser, gc.Equals, test.expectUser)
-		c.Assert(foundKey, gc.DeepEquals, test.expectKey)
+		c.Assert(username, gc.Equals, test.expectUser)
+		c.Assert(key, gc.DeepEquals, test.expectKey)
+	}
+}
+
+var pathCmpTests = []struct {
```

Comment:
I think I'd prefer to have a slightly higher level test than this (possibly in addition) which focused more on the external behaviour, because there's quite a bit of logic that happens before we ever get to pathCmp and it would be nice to be able to change the internal algorithm without changing the tests much.